orttraining/orttraining/python/training/ortmodule/_utils.py

```diff
@@ -104,3 +105,139 @@ def _create_iobinding(io_binding, inputs, model, device):
 
     for value_info in model.graph.output:
         io_binding.bind_output(value_info.name, device.type, device_id=get_device_index(device))
+
+def aten_op_executor_cpp_extension_run_once(f):
+    """
+    Decorator to run a function only once.
+    :param f: function to be run only once during execution time despite the number of calls
+    :return: The original function with the params passed to it if it hasn't already been run before
+    """
+    def aten_op_executor_cpp_extension_wrapper(*args, **kwargs):
+        if not aten_op_executor_cpp_extension_wrapper.has_run:
+            with aten_op_executor_cpp_extension_wrapper.lock:
+                if not aten_op_executor_cpp_extension_wrapper.has_run:
```

Comment:
Do we have to check twice, before and after the lock? How about locking first and then checking? would save one check